lib/dora-parser/src/parser.rs

```diff
@@ -618,23 +638,33 @@ impl<'a> Parser<'a> {
                               mut parse: F)
                               -> Result<Vec<R>, MsgWithPos>
         where F: FnMut(&mut Parser) -> Result<R, MsgWithPos>
+    {
+        self.parse_list(stop, TokenKind::Comma, parse)
+    }
+
+    fn parse_list<F, R>(&mut self,
```

Comment:
Could you instead write a `parse_path` function that returns `Vec<Name>` that doesn't use `parse_list`? I think it would be better to not use parse_list. Right now parse_list can use the `;` terminator, but with closed modules like you proposed we would have a second separator `{`.

Reply:
My intention was to cleanly separate modules-as-packages (open modules) and modules-as-holders-of-static-members (closed modules) from the get-go. Basically,

- an open module
  - can have multiple paths like `foo::bar::baz`
  - can only appear as the first element
  - needs to have its name start with a lowercase letter
  - ends only with `;`, it cannot have `{ ... }`

- a closed module
  - only ever has a single name, e. g. `Foo`, never a path with multiple elements
  - can appear everywhere
  - needs to have its name start with an uppercase letter
  - may end with `;`, or can contain members with `{ ... }`

I think there should never be a case where we can't decide upfront whether we are parsing an open or a closed method, so I'd just use a different `parse_...` method to parse closed modules.

Reply:
IIRC you mentioned LSP. I think we need parsing to be quite forgiving that this works well. E.g. open module might not be the first statement in a file. As soon as this is the case parsing would fail and not show any other errors.

You can still return either ModuleDecl or Module from `parse_module` depending on what form was used. Just because the AST can parse something like `module a::b {}` doesn't mean it's valid code, we can still report an error in semck.

Reply:
That's a good point, I think I will need to think for deeply about this. Thanks!